### PR TITLE
Vmem rework

### PR DESCRIPTION
**Pull request by anton-bondarev**
* Rework vmem subsystem
* Enable x86 vmem
* Enable ARM vmem

### CODE REVIEW

File: src/mem/vmem/mapper.c [hunk 1]
```diff
@@ -0,0 +1,165 @@
+/**
+ * @file
+ * @brief
+ *
+ * @date 05.10.2012
+ * @author Anton Bulychev
```
Comment (Deryugin) [inline src/mem/vmem/mapper.c]: Wrong date/author?

File: src/mem/vmem/vmem_alloc/vmem_alloc_pool.c [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/**
+ * @file
+ * @brief
+ *
+ * @date 24.09.2012
+ * @author Anton Bulychev
```
Comment (Deryugin) [inline src/mem/vmem/vmem_alloc/vmem_alloc_pool.c]: Wrong date/author

File: src/mem/vmem/vmem_alloc/vmem_alloc_single_pool.c [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/**
+ * @file
+ * @brief
+ *
+ * @date 24.09.2012
+ * @author Anton Bulychev
```
Comment (Deryugin) [inline src/mem/vmem/vmem_alloc/vmem_alloc_single_pool.c]: Wrong date/author?

File: src/mem/vmem/mapper.c [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		entry->table[i] = mmu_value(i - 1, entry->table[i - 1] + entry->idx[i - 1]);
+		if (entry->table[i] == NULL) {
+			break;
+		}
+	}
+
+	return entry;
+}
+
+static struct mmu_entry *vmem_entry_from_vaddr( mmu_ctx_t ctx, mmu_vaddr_t virt_addr, struct mmu_entry *entry) {
```
Comment (Deryugin) [inline src/mem/vmem/mapper.c]: `...vaddr( mmu...`->`...vaddr(mmu_...`

File: src/arch/generic/mmu/mmu.c [hunk 5]
```diff
@@ -5,103 +5,146 @@
...
+	case 1:
+		mmu_set_val((unsigned long *) entry,
+			(MMU_ET_PTD | (((unsigned long) value) >> 4)));
+		return;
+	case 2:
+		mmu_set_val((unsigned long *) entry, ((value >> 4) & MMU_PTE_PMASK) | MMU_ET_PTE);
+		return;
+	}
+}
+#if 0
```
Comment (Deryugin) [inline src/arch/generic/mmu/mmu.c]: Remove this dead code?

File: src/mem/vmem/vmem.h [hunk 6]
```diff
@@ -38,32 +38,44 @@ static inline vmem_page_flags_t prot_to_vmem_flags(uint32_t flags) {
...
 extern void vmem_on(void);
 extern void vmem_off(void);
 extern int vmem_mmu_enabled(void);
-extern void mmu_flush_tlb(void);
 
-extern void vmem_get_idx_from_vaddr(mmu_vaddr_t virt_addr, size_t *pgd_idx, size_t *pmd_idx, size_t *pte_idx);
+#define MMU_LAST_LEVEL (MMU_LEVELS - 1)
+
+#ifndef __MMU_SHIFT_1
+#define __MMU_SHIFT_1 (0 * 8 * sizeof(uintptr_t))
```
Comment (Deryugin) [inline src/mem/vmem/vmem.h]: Why not just simply `#define __MMU_SHIFT_1 0`?

File: src/mem/vmem/vmem_alloc/vmem_alloc_pool.c [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+#include <string.h>
+
+#include <mem/page.h>
+
+#include <mem/vmem.h>
+#include <mem/vmem/vmem_alloc.h>
+
+#include <framework/mod/options.h>
+
+static char lvl0_raw[(1024 + 1) * MMU_SIZE(0)] __attribute__ ((aligned(MMU_SIZE(0))));
```
Comment (Deryugin) [inline src/mem/vmem/vmem_alloc/vmem_alloc_pool.c]: Why `1024 + 1`?

File: src/mem/vmem/vmem_nommu.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+#define VMEM_PAGE_WRITABLE    (1 << 0)
+#define VMEM_PAGE_CACHEABLE   (1 << 1)
+#define VMEM_PAGE_USERMODE    (1 << 2)
+/* Used on sparc LEON3 */
+#define VMEM_PAGE_EXECUTABLE  (1 << 3)
+
+typedef uint32_t vmem_page_flags_t;
+
+static inline vmem_page_flags_t prot_to_vmem_flags(uint32_t flags) {
```
Comment (Deryugin) [inline src/mem/vmem/vmem_nommu.h]: This function is duplicated in vmem.h

Comment (anton-bondarev) [reply src/mem/vmem/vmem_nommu.h]: Yes, but I suggest to fix it in the  future

File: src/mem/vmem/vmem_alloc/vmem_alloc_pool.c (hunk 7 above)
Comment (anton-bondarev) [reply src/mem/vmem/vmem_alloc/vmem_alloc_pool.c]: done

File: src/mem/vmem/vmem.h (hunk 6 above)
Comment (anton-bondarev) [reply src/mem/vmem/vmem.h]: done

File: src/mem/vmem/mapper.c [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	vmem_entry_get_idxs(ctx, virt_addr, &entries);
+	vmem_entry_get_tables(ctx, virt_addr, &entries);
+
+	vmem_set_pte_flags(entry->table[MMU_LAST_LEVEL] + entry->idx[MMU_LAST_LEVEL], 0);
+	vmem_set_pte_flags(entry->table[MMU_LAST_LEVEL] + entry->idx[MMU_LAST_LEVEL], flags);
+	return 0;
+}
+
+/**
+ * @brief Change PTE attributes for given memory range: [virt_addr, virt_addr + len]
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/mem/vmem/mapper.c]: done

File: src/arch/arm/mmu/mmu_small_page.c [hunk 10]
```diff
@@ -13,69 +13,62 @@
...
 }
 
-void mmu_pte_unset(mmu_pgd_t *pte) {
-	*pte = 0x0;
-	dcache_flush(pte, sizeof(*pte));
-}
+void mmu_set(int lvl, uintptr_t *entry, uintptr_t value) {
+	switch (lvl) {
+	case 0:
+		*entry = (value & ~0x3FF) | ARM_MMU_TYPE_PAGE;
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/arch/arm/mmu/mmu_small_page.c]: done

File: src/arch/arm/mmu/mmu_small_page.c [hunk 11]
```diff
@@ -13,69 +13,62 @@
...
-void mmu_pgd_unset(mmu_pgd_t *pgd) {
-	*pgd = 0x0;
-}
-
-void mmu_pmd_unset(mmu_pgd_t *pmd) {
-	*pmd = 0x0;
+uintptr_t *mmu_get(int lvl, uintptr_t *entry) {
+	switch (lvl) {
+	case 0:
+		return (uintptr_t *) (*entry & ~0x3FF);
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/arch/arm/mmu/mmu_small_page.c]: done

File: src/arch/microblaze/mmu/mmu.c [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	case 1:
+		return (uintptr_t *) ((((unsigned long) *entry) & MMU_PTD_PMASK) << 4);
+	case 2:
+		return (uintptr_t *)(((*entry) & MMU_PTE_PMASK) << 4);
+	}
+	return 0;
+}
+
+/* Present functions */
+int mmu_present(int lvl, uintptr_t *entry)  {
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/arch/microblaze/mmu/mmu.c]: done

File: src/arch/microblaze/mmu/mmu.c [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+	case 1:
+		return *entry & MMU_ET_PTD;
+	case 2:
+		return *entry & MMU_ET_PTE;
+	}
+	return 0;
+}
+
+/* Unset functions */
+void mmu_unset(int lvl, uintptr_t *entry)  {
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/arch/microblaze/mmu/mmu.c]: done

File: src/mem/vmem/mapper.c (hunk 4 above)
Comment (anton-bondarev) [reply src/mem/vmem/mapper.c]: done

**Pull request merged**